src/main/java/com/aws/greengrass/clientdevices/auth/iot/GreengrassV2DataClientFactory.java

```diff
@@ -0,0 +1,76 @@
+import com.aws.greengrass.util.Utils;
+import software.amazon.awssdk.auth.credentials.AnonymousCredentialsProvider;
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClient;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClientBuilder;
+
+import java.net.URI;
+import javax.inject.Inject;
+
+/**
+ * Provides GG service clients built specifically for CDA use case.
+ * Ideally this factory should maintain the lifecycle of its clients,
+ * but for now its upto the consumer of the clients to maintain their lifecycle.
+ */
+public class GreengrassV2DataClientFactory {
+    private final GreengrassServiceClientFactory greengrassServiceClientFactory;
+
+    /**
+     * Construct Greengrass Client Factory.
+     *
+     * @param greengrassServiceClientFactory GG Service Client Factory to retrieve the latest device-config
+     */
+    @Inject
+    public GreengrassV2DataClientFactory(GreengrassServiceClientFactory greengrassServiceClientFactory) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+    }
+
+    /**
+     * Provides a new GG v2 Data client without implicit retry policy.
+     * Consumer should handle retries when appropriate and maintain the lifecycle of the client.
+     * The Nucleus provided V2DataClient has a built in retry policy which cannot be overridden.
+     *
+     * @return GreengrassV2DataClient
+     */
+    public GreengrassV2DataClient getClient()  {
+        DeviceConfiguration deviceConfiguration = greengrassServiceClientFactory.getDeviceConfiguration();
+        String awsRegion = getAwsRegion(deviceConfiguration);
+        String ggServiceEndpoint = ClientConfigurationUtils.getGreengrassServiceEndpoint(deviceConfiguration);
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        GreengrassV2DataClientBuilder clientBuilder = GreengrassV2DataClient.builder()
+                .credentialsProvider(AnonymousCredentialsProvider.create())
+                .httpClient(httpClient.build())
+                .overrideConfiguration(ClientOverrideConfiguration.builder()
+                        .retryPolicy(RetryPolicy.none()).build());
+
+        clientBuilder.region(Region.of(awsRegion));
+        clientBuilder.endpointOverride(URI.create(ggServiceEndpoint));
+        return clientBuilder.build();
+    }
+
+    private String getAwsRegion(DeviceConfiguration deviceConfiguration) {
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        if (Utils.isEmpty(awsRegion)) {
+            throw new RuntimeException("Failed to retrieve Greengrass service endpoint. "
```

Comment:
why are we throwing unchecked exceptions? Who is handling this, and what are they doing about it?

Reply:
Just replicating Nucleus's behavior since these are similar critical errors and CDA cannot continue with it: [InvalidEnvironmentStageException](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/8a8ff7f53eac7a301103969fcbbb2d567e3ccdf8/src/main/java/com/aws/greengrass/componentmanager/ClientConfigurationUtils.java#L54-L57), [link](https://github.com/aws-greengrass/aws-greengrass-nucleus/blob/dd8e059e802f44a763819a01e2088263ae56f3b3/src/main/java/com/aws/greengrass/easysetup/GreengrassSetup.java#L303-L307). Do you have a suggestion on what exception this should be? [InvalidConfigurationException](https://github.com/aws-greengrass/aws-greengrass-client-device-auth/blob/main/src/main/java/com/aws/greengrass/clientdevices/auth/exception/InvalidConfigurationException.java) ?

Reply:
Yes invalid configuration would be fine. I'm not too worried about the exception itself, I just really want to make sure it is handled appropriately.

Reply:
We aren't catching runtime exceptions anywhere. Please do not throw them

Reply:
fixed it